### DISCUSSION TITLE
USB audio audio_4_channel_mic H7 produce only one channel

### DISCUSSION THREAD

**Discussion by Lurcy38** [hathach/tinyusb#2290] [General]
Hello , 
I must say, I am beginner in tinyusb. I started the tinyUsb evaluation because we have a project that need audio USB on stm32h7 and stm32h5.
We are ready to help for the H5 porting if needed.
The first step is to build demos and check the API and if the solution is ok. audio_4_channel_mic and uac2_headset is very close of to our use case. 
So, I built theses demos on  H7 (H5 is not supported yet) 
make BOARD=stm32h743eval get-deps
make BOARD=stm32h743eval all 
the build is ok and perfect, but the result is confusing.
I have just amplified the signal to be more visible …
When I use audacity, to check the result, is see only one channel with the signal. whith plot_audio_samples.py, I see also a single channel. 

I use the release [0.15.0](https://github.com/hathach/tinyusb/releases/tag/0.15.0)  and the master . both have the same result?

How to fix this issue and get a stereo audio usb input ?
Where am I wrong?

any help is welcome :)

regards

**Comment by HiFiPhile** [comment]
With #2259 and #2282 audio_4_channel_mic should working. What you saw with `plot_audio_samples.py` ?
Did you made a clean build after pulled master branch ?

**Comment by Lurcy38** [comment]
hello,

here are screenshots for audacity and plot_audio_samples.py
https://postimg.cc/620w7q6t 
https://postimg.cc/tZ0t1RSS

i just modified the signal generation  to render  a larger signal on audacity

int cumul = 0;
bool tud_audio_tx_done_post_load_cb(uint8_t rhport, uint16_t n_bytes_copied, uint8_t itf, uint8_t ep_in, uint8_t cur_alt_setting)
{
  (void) rhport;
  (void) n_bytes_copied;
  (void) itf;
  (void) ep_in;
  (void) cur_alt_setting;

  uint16_t dataVal=cumul;

  // Generate dummy data
  for (uint16_t cnt = 0; cnt < CFG_TUD_AUDIO_FUNC_1_N_TX_SUPP_SW_FIFO; cnt++)
  {
    uint16_t * p_buff = i2s_dummy_buffer[cnt];              // 2 bytes per sample
    for (uint16_t cnt2 = 0; cnt2 < AUDIO_SAMPLE_RATE/1000; cnt2++)
    {
      for (uint8_t cnt3 = 0; cnt3 < CFG_TUD_AUDIO_FUNC_1_CHANNEL_PER_FIFO_TX; cnt3++)
      {
        *p_buff++ = dataVal;
      }
      dataVal++;
    }
  }
  cumul = dataVal;
  return true;
}

**Reply by HiFiPhile** [reply]
From your posted code you didn't update to master branch correctly. These are old code from [0.15.0](https://github.com/hathach/tinyusb/releases/tag/0.15.0)

**Reply by Lurcy38** [reply]
ok the latest release is KO, but moving on the master  is perfect  for the audio 4 mics, I only changed  the python line
        device = 'Microphone (MicNode_4_Ch), Windows WDM-KS' that down't work on windows10 and python 3.8

many thanks for your support :)

**Reply by HiFiPhile** [reply]
> Windows WDM-KS

Yeah Windows doesn't play well with sounddevice, sometimes WASAPI works other time need WDM-KS.

**Reply by Lurcy38** [reply]
Hello phil
Sorry to bother you again but I need some advice concerning the USB IP porting to make a decision.
I work at  STMicroelectronics GNB  in the MCU division. We plan to use  tinyUSB in our project, because we need some usb class already present in tinyusb and not in our internal solution. 
I validated that the H7 support works fine without audio pops , but we need also the H5 not present in your board support. 
ST has changed its USB IP for H5 and future SOC , it is no longer a synopsis IP, but an internal IP. We have already a low-level support for this IP called stm32h5xx_ll_usb. This module takes in charge basic USB operation such as fifo TX/RX ep etc….
The question is:  From your knowledge, how is difficult to port the equivalent to dcd_synopsys.c for the new IP ST ?
The dcd_synopsys.c code is completely based on the datasheet ~800 active lines (no  ll API support). 
From the information I gave you, how full-time weeks for an MCU experimented Personne to get the driver functional? 
Regards

**Reply by HiFiPhile** [reply]
> it is no longer a synopsis IP, but an internal IP

From the works in #2203, seems it's the same IP as G0 ?

That IP is supported in https://github.com/hathach/tinyusb/blob/master/src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c

>  how full-time weeks for an MCU experimented Personne to get the driver functional

It's hard to say, last time I spent 2 weeks spare time to write Microchip SAM7X driver.

**Reply by Lurcy38** [reply]
Thank a lot for details.  the  H5 and L5 shares the same IP. An evolution of the G0.
i have now all info i need..., again many thanks
regards
